accord-core/src/main/java/accord/primitives/Deps.java

```diff
@@ -0,0 +1,978 @@
+                        break;
+                    }
+                    else if (r < right[rk])
+                    {
+                        o += right[rk] - r;
+                        r = right[rk];
+                    }
+
+                    assert o == r && ok == rk && right[ok] == o;
+                    ok++;
+                    rk++;
+                    lk++;
+                }
+            }
+
+            if (out == null)
+                return that;
+        }
+        else
+        {
+            out = new int[left.length + right.length];
+        }
+
+        while (lk < leftKeys.size() && rk < rightKeys.size())
+        {
+            int ck = leftKeys.get(lk).compareTo(rightKeys.get(rk));
+            if (ck < 0)
+            {
+                while (l < left[lk])
+                    out[o++] = remap(left[l++], remapLeft);
+                out[ok++] = o;
+                lk++;
+            }
+            else if (ck > 0)
+            {
+                while (r < right[rk])
+                    out[o++] = remap(right[r++], remapRight);
+                out[ok++] = o;
+                rk++;
+            }
+            else
+            {
+                while (l < left[lk] && r < right[rk])
+                {
+                    int nextLeft = remap(left[l], remapLeft);
+                    int nextRight = remap(right[r], remapRight);
+
+                    if (nextLeft <= nextRight)
+                    {
+                        out[o++] = nextLeft;
+                        l += 1;
+                        r += nextLeft == nextRight ? 1 : 0;
+                    }
+                    else
+                    {
+                        out[o++] = nextRight;
+                        ++r;
+                    }
+                }
+
+                while (l < left[lk])
+                    out[o++] = remap(left[l++], remapLeft);
+
+                while (r < right[rk])
+                    out[o++] = remap(right[r++], remapRight);
+
+                out[ok++] = o;
+                rk++;
+                lk++;
+            }
+        }
+
+        while (lk < leftKeys.size())
+        {
+            while (l < left[lk])
+                out[o++] = remap(left[l++], remapLeft);
+            out[ok++] = o;
+            lk++;
+        }
+
+        while (rk < rightKeys.size())
+        {
+            while (r < right[rk])
+                out[o++] = remap(right[r++], remapRight);
+            out[ok++] = o;
+            rk++;
+        }
+
+        if (o < out.length)
+            out = Arrays.copyOf(out, o);
+
+        return new Deps(keys, txnIds, out);
+    }
+
+    private static int[] copy(int[] src, int to, int length)
+    {
+        if (length == 0)
+            return NO_INTS;
+
+        int[] result = new int[length];
+        System.arraycopy(src, 0, result, 0, to);
+        return result;
+    }
+
+    // TODO: optimise for case where none removed
+    public Deps without(Predicate<TxnId> remove)
```

Comment:
ignoring `merge` method for now, so this is the last one for this review!  I pushed tests and it is failing for me, trying to figure out why but what I see is the following: remove T1 (which exists) and check key -> [T] to see if still present, and check someKeys(T) -> [K] as well... what I am seeing is that a different T2 gets removed instead....  but T2 is known in the txnId array, its just that the offsets are messed up for this txn only

```
accord.primitives.Deps without = deps.test.without(i -> i.equals(txnId));
```